src/interface/misc.rs

```diff
@@ -381,70 +381,58 @@ pub struct RustSemaphore {
 impl RustSemaphore {
     pub fn new(value_handle: u32, is_shared: bool) -> Self {
         Self {
-            value: RustAtomicU32::new(value_handle),
+            value: Mutex::new(value_handle),
             is_shared: RustAtomicBool::new(is_shared),
         }
     }
 
-    pub fn lock(&self) -> bool{
-        while self.value.load(RustAtomicOrdering::Relaxed) == 0 { interface::lind_yield(); }
-
-        let result = self.value.fetch_update(RustAtomicOrdering::Relaxed, RustAtomicOrdering::Relaxed, |x| {
-            if x > 0 { Some(x - 1) } else { Some(0) }
-        });
-
-        match result {
-            Ok(_) => true,
-            Err(_) => false,
+    pub fn lock(&self) {
+        loop {
+            let mut value = self.value.lock();
+            if *value == 0 {
+                interface::lind_yield();
+                continue;
+            }
+                
+            *value = if *value > 0 { *value - 1 } else { 0 };
```

Comment:
this should be an if/else clause, if you do it that way you dont need the continue